src/Storages/Rocksdb/StorageEmbeddedRocksdb.cpp

```diff
@@ -0,0 +1,427 @@
+};
+
+StorageEmbeddedRocksdb::StorageEmbeddedRocksdb(const StorageID & table_id_,
+        const String & relative_data_path_,
+        const StorageInMemoryMetadata & metadata_,
+        bool attach,
+        Context & context_,
+        const String & primary_key_)
+    : IStorage(table_id_), primary_key{primary_key_}
+{
+    setInMemoryMetadata(metadata_);
+    rocksdb_dir = context_.getPath() + relative_data_path_ + "/rocksdb";
+    if (!attach)
+    {
+        Poco::File(rocksdb_dir).createDirectories();
+    }
+    initDb();
+}
+
+void StorageEmbeddedRocksdb::truncate(const ASTPtr &, const StorageMetadataPtr & , const Context &, TableExclusiveLockHolder &)
+{
+    rocksdb_ptr->Close();
+    Poco::File(rocksdb_dir).remove(true);
+    Poco::File(rocksdb_dir).createDirectories();
+    initDb();
+}
+
+void StorageEmbeddedRocksdb::initDb()
+{
+    rocksdb::Options options;
+    rocksdb::DB * db;
+    options.create_if_missing = true;
+    rocksdb::Status status = rocksdb::DB::Open(options, rocksdb_dir, &db);
+
+    if (status != rocksdb::Status::OK())
+        throw Exception("Fail to open rocksdb path at: " +  rocksdb_dir, ErrorCodes::SYSTEM_ERROR);
+    rocksdb_ptr = std::unique_ptr<rocksdb::DB>(db);
+}
+
+
+Pipe StorageEmbeddedRocksdb::read(
+        const Names & column_names,
+        const StorageMetadataPtr & metadata_snapshot,
+        const SelectQueryInfo & query_info,
+        const Context & /*context*/,
+        QueryProcessingStage::Enum /*processed_stage*/,
+        size_t /*max_block_size*/,
+        unsigned num_streams)
+{
+    metadata_snapshot->check(column_names, getVirtuals(), getStorageID());
+    Block sample_block = metadata_snapshot->getSampleBlock();
+    size_t primary_key_pos = sample_block.getPositionByName(primary_key);
+
+    FieldVector keys;
+    bool all_scan = false;
+
+    std::tie(keys, all_scan) = getFilterKeys(primary_key, query_info);
+    if (all_scan)
+    {
+        MutableColumns columns = sample_block.cloneEmptyColumns();
+        auto it = std::unique_ptr<rocksdb::Iterator>(rocksdb_ptr->NewIterator(rocksdb::ReadOptions()));
+        for (it->SeekToFirst(); it->Valid(); it->Next())
+        {
+            ReadBufferFromString key_buffer(it->key());
+            ReadBufferFromString value_buffer(it->value());
+
+            for (const auto [idx, column_type] : ext::enumerate(sample_block.getColumnsWithTypeAndName()))
+            {
+                column_type.type->deserializeBinary(*columns[idx], idx == primary_key_pos? key_buffer: value_buffer);
+            }
+        }
+
+        if (!it->status().ok())
+        {
+            throw Exception("Engine " + getName() + " got error while seeking key value datas: " + it->status().ToString(),
+                ErrorCodes::LOGICAL_ERROR);
+        }
+        UInt64 num_rows = columns.at(0)->size();
+        Chunk chunk(std::move(columns), num_rows);
+        return Pipe(std::make_shared<SourceFromSingleChunk>(sample_block, std::move(chunk)));
```

Comment:
Full scan code path still requires loading all the data into memory, it will be great if it can stream the data as the code path above.
That will not only solve the memory issue, but also give some progress to the user